p2p/messenger.go

```diff
@@ -0,0 +1,342 @@
+package p2p
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/centrifuge/go-centrifuge/errors"
+
+	"github.com/libp2p/go-libp2p-host"
+	"github.com/libp2p/go-libp2p-protocol"
+
+	pb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/protocol"
+	ggio "github.com/gogo/protobuf/io"
+	"github.com/jbenet/go-context/io"
+	inet "github.com/libp2p/go-libp2p-net"
+	"github.com/libp2p/go-libp2p-peer"
+)
+
+// ErrReadTimeout must be used when receiving timeout while reading
+const ErrReadTimeout = errors.Error("timed out reading response")
+
+// ErrInvalidatedMessageSender must be used when the message sender object created is no longer valid (connection has dropped)
+const ErrInvalidatedMessageSender = errors.Error("message sender has been invalidated")
+
+type bufferedWriteCloser interface {
+	ggio.WriteCloser
+	Flush() error
+}
+
+// The Protobuf writer performs multiple small writes when writing a message.
+// We need to buffer those writes, to make sure that we're not sending a new
+// packet for every single write.
+type bufferedDelimitedWriter struct {
+	*bufio.Writer
+	ggio.WriteCloser
+}
+
+func newBufferedDelimitedWriter(str io.Writer) bufferedWriteCloser {
+	w := bufio.NewWriter(str)
+	return &bufferedDelimitedWriter{
+		Writer:      w,
+		WriteCloser: ggio.NewDelimitedWriter(w),
+	}
+}
+
+func (w *bufferedDelimitedWriter) Flush() error {
+	return w.Writer.Flush()
+}
+
+type p2pMessenger struct {
+	host host.Host // the network services we need
+	self peer.ID   // Local peer (yourself)
+
+	timout time.Duration
+	ctx    context.Context
+
+	strmap map[peer.ID]map[protocol.ID]*messageSender
+	smlk   sync.Mutex
+
+	plk sync.Mutex
+
+	msgHandlers map[pb.MessageType]func(ctx context.Context, peer peer.ID, protoc protocol.ID, msg *pb.P2PEnvelope) (*pb.P2PEnvelope, error)
+}
+
+func newP2PMessenger(ctx context.Context, host host.Host, p2pTimeout time.Duration) *p2pMessenger {
+	return &p2pMessenger{
+		ctx:         ctx,
+		host:        host,
+		self:        host.ID(),
+		timout:      p2pTimeout,
+		strmap:      make(map[peer.ID]map[protocol.ID]*messageSender),
```

Comment:
I believe keeping streams open is interesting but uncertain when scaling up. I like the concept, but I would  have loved to simplify this approach at the moment. Let's keep it now that is implemented.